Add terraform module for azure auto discovery

avatus opened the pull request:
This will add a terraform module that will support the automatic discovery of VMs in azure, using an OIDC integration. 

Closes https://github.com/gravitational/teleport/issues/60818

## Manual Test Plan
### Test Environment
Cloud staging environment
Azure Account with VMs

### Test Cases
- [ ] Test module with Azure VM matcher
- [ ] Ensure Azure VMs were discovered

Changelog: Added a terraform module to support automatic discovery of Azure resources.

avatus committed:
Add terraform module for azure auto discovery

This will add a terraform module that will support the automatic
discovery of VMs in azure, using an OIDC integration.

Closes #60818

Changelog: Added a terraform module to support automatic discovery of
azure resources.

avatus commented on examples/terraform-modules/teleport-azure-discovery/main.tf:
```diff
@@ -0,0 +1,164 @@
+    }
+    teleport = {
+      source  = "terraform.releases.teleport.dev/gravitational/teleport"
+      version = "~> 18.5"
```
I believe the required resources are in v18 now, but i think youire right ath 18.5 might be too old. might have to do a patch or just do 18.6

avatus commented on examples/terraform-modules/teleport-azure-discovery/main.tf:
```diff
@@ -0,0 +1,164 @@
+  discovery_resource_groups = var.discovery_resource_group_names
+  proxy_addr                = var.proxy_addr
+  discovery_group           = var.discovery_group_name
+  identity_resource_group   = var.identity_resource_group_name != "" ? var.identity_resource_group_name : local.discovery_resource_groups[0]
```
yeah i will. this might also just be vestigial so ill see if i can remove it, and if not, will comment

greedy52 replied:
use [coalesce](https://developer.hashicorp.com/terraform/language/functions/coalesce) ?

hugoShaka reviewed (commented):
I'm no Terraform expert but this looks ok.

Can we make sure that:
- this code is linted with `tf lint` and the CI enforces it
- there is a clear owner of the module and it is covered by the test plan

We had other TF modules in the past that went unmainted and I had to remove. My main concern is that this does not follow the same path. I'd like us to ensure the same quality/support level as the go code we write.

greedy52 commented on examples/terraform-modules/teleport-azure-discovery/examples/single-subscription/main.tf:
```diff
@@ -0,0 +1,20 @@
+module "teleport_azure_discovery" {
+  source = "../.."
+
+  subscription_id                = var.subscription_id
```
should we support multiple subscriptions or even wildcard?

avatus replied:
as far as multiple subscruption IDs goes, im unsure how it would work atm. perhaps we go with single sub now and then expand when we do multi-account stuff with azure?

ill look into wildcards for regions/resource groups

greedy52 commented on examples/terraform-modules/teleport-azure-discovery/examples/single-subscription/main.tf:
```diff
@@ -0,0 +1,20 @@
+  identity_resource_group_name = var.identity_resource_group_name
+  tags                         = var.tags
+  discovery_group_name         = var.discovery_group_name
+  installer_script_name        = var.installer_script_name
```
@GavinFrazar we should be consistent with naming across modules:
- https://github.com/gravitational/teleport/pull/62106

i quite like how Gavin groups the variable names with a prefix, but up to you guys.

(also not sure what's the more common `prefix` vs `name_name`)

GavinFrazar committed:
Add license

GavinFrazar committed:
Refactor module

- align variable names
- conditional creation
- remove all inputs from the example

GavinFrazar committed:
Move Azure module into published TF modules dir

greedy52 reviewed (approved):
any plan to support non-integration discovery?

GavinFrazar committed:
Fix azurerm_federated_identity_credential

Remove unsupported argument "resource_group_name".

greedy52 commented:
could we provide a manual test plan for this PR? @GavinFrazar do you prefer updating this one or maybe opening a new PR as you?

GavinFrazar committed:
Remove role assignment output

It's not useful to output this.

GavinFrazar committed:
Sort role definition permissions

GavinFrazar committed:
Add subscriptions/read permission to oidc role

This permission will be used by:
1. discovery service to perform wildcard subscription discovery
2. auth service to perform wildcard subscription VM join validation

GavinFrazar committed:
Simplify client config data source

It's so cheap to read the data source that we don't need to make it
conditional and it improves the terraform plan.

GavinFrazar committed:
Add support to match wildcard Azure subscriptions

In order to discover VMs in all subscriptions, the module must allow the
user to configure the role and role assignment scopes.

Additionally, the module must allow the user to specify the
subscription(s) which should be discovered.

GavinFrazar committed:
make docs

alexhemard committed:
add azure discovery module to published tf modules list

alexhemard committed:
[terraform] update azure discovery module matcher to match aws

alexhemard commented:
@codex review

alexhemard commented:
eb32cc9afcd61190929436418d49148ec6db6e1c
- Updated matchers to `azure_matchers` to match aws discovery module (https://github.com/gravitational/teleport/pull/65002)
- at least one subscription matcher is required
- creates a role definition + assignment per scope, with one scope per subscription matcher by default
- added precondition to require `azure_role_assignment_scopes` when using a wildcard subscription matcher

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/azure_role_definition.tf:
```diff
@@ -0,0 +1,51 @@
+  ))
+}
+
+# Custom role for discovery permissions for each scope.
+resource "azurerm_role_definition" "teleport_discovery" {
+  for_each = local.create ? toset(local.azure_role_assignment_scopes) : toset([])
+
+  assignable_scopes = [each.value]
+  description       = "Azure role that allows a Teleport Discovery Service to discover VMs."
+  # Split each scope by '/' and hyphenate the last two segments
+  # e.g. "subscriptions-{uuid}", "managementGroups-{name}".
+  name  = "${var.azure_role_definition_name}-${join("-", slice(split("/", each.value), length(split("/", each.value)) - 2, length(split("/", each.value))))}"
+  scope = each.value
```
I think we can create just one role definition with many assignable scopes. I'm not sure it even matters which scope we choose to create the definition at

alexhemard replied:
changed to 1 role definition - n assignments. confirmed it doesn't matter which scope `azurerm_role_definition` has

alexhemard committed:
feedback, one role definition, n assignments

alexhemard commented on integrations/terraform-modules/teleport/discovery/azure/teleport_provision_token.tf:
```diff
@@ -0,0 +1,32 @@
+  spec = {
+    azure = {
+      allow = [for sub in local.azure_matcher_subscriptions : {
+        subscription = sub
```
seems valid, also the join token will be based on tenant id, not subscription wildcard 

per https://github.com/gravitational/teleport/pull/65866

GavinFrazar replied:
yeah since we opted to allow tenant ID in tokens instead of expanding a wildcard subscription, I would check for wildcard like this:

```suggestion
      allow = (
        contains(local.match_azure_subscriptions, "*")
        ? [{ tenant = local.azure_tenant_id }]
        : [for sub in var.match_azure_subscriptions : { subscription = sub }]
      )
```

alexhemard replied:
@GavinFrazar are we good to make this change since https://github.com/gravitational/teleport/pull/65866 hasn't been merged yet?

alexhemard replied:
Also _should_ we assume using the terraform provider client's tenant ID? 

currently forcing `var.azure_role_assignment_scopes` to be set to use wildcard subscriptions; maybe we should do something similar for the join token?

See https://github.com/gravitational/teleport/pull/62145/changes#diff-a237cc48f0bdfc2c27715c3442ad75addfc50c143934f1b2b51a4ee16a5b98c9R52-R58

alexhemard commented on integrations/terraform-modules/teleport/discovery/azure/outputs.tf:
```diff
@@ -0,0 +1,24 @@
+output "azure_discovery_role_definition" {
+  description = "The Azure role definition for the Teleport Discovery Service."
+  value       = one(azurerm_role_definition.teleport_discovery[*])
```
seems fine here for the output? 

`If the collection is empty, one returns null` [[docs]](https://developer.hashicorp.com/terraform/language/functions/one)

GavinFrazar replied:
This is a hallucination about the behavior of the `one()` function in terraform.
I've seen codex make the same mistake, not sure why exactly but LLMs seem to think if there are 0 elements in the list that it will error out, when in fact it will just be null as you also pointed out 🤷‍♂️

alexhemard committed:
allow rule for rg, subscription

alexhemard committed:
add provision token allow rules, provision token create var

r0mant reviewed (commented):
@alexhemard @GavinFrazar I'll echo @greedy52's comment: https://github.com/gravitational/teleport/pull/62145#pullrequestreview-3707505149, and the same comment I left @marcoandredinis on his PR: https://github.com/gravitational/teleport/pull/63502#pullrequestreview-4123436162.

Does this module support disabling OIDC auth to support self-hosted scenarios where users just give their discovery services a managed identity with proper permissions?

alexhemard commented:
> Does this module support disabling OIDC auth to support self-hosted scenarios where users just give their discovery services a managed identity with proper permissions?

I'm not sure how much value this module provides in this use case  - or if the Azure provider is needed at all.

If they already have a managed identity for the VMs running the discover service only `teleport_discovery_config` and `teleport_provision_token` resources would be needed afaik

@r0mant @GavinFrazar

alexhemard committed:
add create_teleport_integration bool for self-hosted private clusters

alexhemard committed:
add var create_azure_managed_identity

alexhemard commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,230 @@
+variable "teleport_provision_token_allow_rules" {
+  description = "Custom allow rules for the Teleport provision token. Required when using a wildcard (`*`) subscription matcher and `create_teleport_provision_token` is `true`."
+  type = list(object({
+    subscription    = optional(string)
+    resource_groups = optional(list(string))
+    tenant          = optional(string)
```
https://github.com/gravitational/teleport/pull/65866 was merged

alexhemard commented on integrations/terraform-modules/teleport/discovery/azure/teleport_discovery_config.tf:
```diff
@@ -0,0 +1,71 @@
+          join_method = "azure"
+          join_token  = local.teleport_provision_token_name
+          script_name = var.teleport_installer_script_name
+        }
```
Added `create_teleport_integration` var to disable integration creation

alexhemard committed:
misc fixes

- fix suffix issues w/ azure resources
- fix allow rules type mismatch in provision token
- use random resource for suffix

alexhemard committed:
add tests

alexhemard committed:
allow teleport_integration_name to be nullable, tests

alexhemard commented on integrations/terraform-modules/teleport/discovery/azure/azure_role_definition.tf:
```diff
@@ -0,0 +1,55 @@
+################################################################################
+
+locals {
+  azure_role_assignment_scopes = coalescelist(
+    var.azure_role_assignment_scopes,
+    [for sub in local.azure_matcher_subscriptions : "/subscriptions/${sub}"],
```
Addressed

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,251 @@
+      for matcher in var.azure_matchers :
+      !contains(matcher.subscriptions, "*") || length(matcher.subscriptions) == 1
+    ])
+    error_message = "Wildcard ('*') must be the only entry in a matcher's subscriptions list."
```
Why? We can just simplify the list to wildcard

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,251 @@
+  nullable    = false
+}
+
+variable "create_azure_managed_identity" {
+  description = "Whether Azure managed identity and role resources are created (true) or not (false). When false, no Azure resources are created. Must be set to `true` when `create_teleport_integration` is `true`."
+  type        = bool
+  default     = true
+  nullable    = false
+}
```
I think we should always create the managed identity and role definition.

```suggestion
```

alexhemard replied:
this is useful for the ambient credentials use-case

GavinFrazar replied:
using the created identity is optional, so we don't need this toggle for that, but it does allow you to toggle off the creation of a role definition which takes a long time (~5 minutes in my experience)

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,251 @@
+  type        = bool
+  default     = true
+  nullable    = false
+}
```
We should always create the token too

```suggestion
```

alexhemard replied:
I think this is fine to leave out for now

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,251 @@
+}
+
+variable "teleport_integration_name" {
+  description = "Name for the `teleport_integration` resource. Required when `create_teleport_integration` is true. When `create_teleport_integration` is false, set to reference an existing integration or null to authenticate using ambient Azure credentials."
```
Let's avoid the pattern where the module may or may not own the resources it references - it should be one or the other to keep the module simple.

There are 3 branches here:
1. create and use integration
2. use an existing integration
3. don't use an integration

We can simplify by eliminating the "use existing" case.

There should be a bool variable like `use_oidc_integration` and if it's true we create an integration and use it. Otherwise, we don't create an integration and don't reference an integration in the discovery config.

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,251 @@
+  description = "Name for the `teleport_integration` resource. Required when `create_teleport_integration` is true. When `create_teleport_integration` is false, set to reference an existing integration or null to authenticate using ambient Azure credentials."
+  type        = string
+  default     = "discovery"
+  nullable    = true
```
```suggestion
  nullable    = false
```

GavinFrazar commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,251 @@
+  nullable    = true
+}
+
+variable "create_teleport_integration" {
```
```suggestion
variable "use_oidc_integration" {
```

alexhemard committed:
feedback

- rename create_teleport_integration to use_oidc_integration.
- revert nullable teleport_integration_name
- revert create_teleport_provision_token

alexhemard commented on integrations/terraform-modules/teleport/discovery/azure/teleport_provision_token.tf:
```diff
@@ -0,0 +1,58 @@
+  ])
+
+  token_allow_rules = (
+    var.teleport_provision_token_allow_rules != null
+    ? var.teleport_provision_token_allow_rules
+    : local.token_allow_rules_from_matchers
```
seems out of scope

greedy52 commented on docs/pages/reference/infrastructure-as-code/terraform-modules/teleport-discovery-azure/examples/examples.mdx:
```diff
@@ -0,0 +1,15 @@
+---
+title: Teleport AZURE discovery examples
```
`AZURE`-> `Azure`. Please fix the generation script

greedy52 commented on integrations/terraform-modules/teleport/discovery/azure/examples/single-subscription/main.tf:
```diff
@@ -0,0 +1,86 @@
+  teleport_discovery_group_name = "cloud-discovery-group"
+  teleport_proxy_public_addr    = "example.teleport.sh:443"
+
+  # Name of an existing Azure Resource Group where Azure resources will be created.
```
It says "Name of an existing ..." but it was created below. which one is correct? e.g should it use `data` below or fixing the comment here.

alexhemard replied:
It's not created by the module and needs to reference an existing resource group name. I can move the resource group resource before the module so it reads better

alexhemard replied:
Moved the resource group before the module and added a comment that it will be where the Azure resources are created

greedy52 commented on integrations/terraform-modules/teleport/discovery/azure/examples/single-subscription/versions.tf:
```diff
@@ -0,0 +1,20 @@
+
+    teleport = {
+      source  = "terraform.releases.teleport.dev/gravitational/teleport"
+      version = ">= 18.5.1"
```
this version doesn't match the one in the main module.

alexhemard replied:
Needs azure tenant join token support - seems like it should be be `>= 18.7.6` 
per https://github.com/gravitational/teleport/pull/65989

greedy52 commented on integrations/terraform-modules/teleport/discovery/azure/variables.tf:
```diff
@@ -0,0 +1,244 @@
+# Required variables
+################################################################################
+
+variable "azure_managed_identity_location" {
```
are `azure_managed_identity_location` and `azure_resource_group_name` required or optional?

alexhemard replied:
In the description

> `Required when `create_azure_managed_identity` is `true`."`

When the managed identity isn't created, there aren't any azure resources created by the module so this isn't required in that case

greedy52 commented on integrations/terraform-modules/teleport/discovery/azure/teleport_integration.tf:
```diff
@@ -0,0 +1,46 @@
+      condition     = var.create_azure_managed_identity
+      error_message = "use_oidc_integration requires create_azure_managed_identity to be true."
+    }
+  }
```
nit: is this precondition needed? assuming `depends_on` is proper way to depend on something. or can the validation be done in variables.  similar questions to other preconditions.

alexhemard replied:
The OIDC integration requires the Azure managed identity to be created in this module configured by variables. 
- `use_oidc_integration` (creates the teleport integration and federated identity) 
- `create_azure_managed_identity` (creates the azure managed identity and role assignments)

The integration can be disabled and the managed identity can be created (or not) for self-hosted clusters

`depends_on` would probably error if a dependent resource wasn't created and variable validation can't cross reference other variables unless we bump terraform version requirement to `>= 1.9` which is why there's a preconditon for now

greedy52 commented on integrations/terraform-modules/teleport/discovery/azure/azure_role_definition.tf:
```diff
@@ -0,0 +1,55 @@
+locals {
+  azure_role_assignment_scopes = coalescelist(
+    var.azure_role_assignment_scopes,
+    [for sub in local.azure_matcher_subscriptions : "/subscriptions/${sub}"],
```
can `local.azure_matcher_subscriptions` be wildcard here like `/subscriptions/*`?does it work for the resources below? e.g does assignable scope support wildcard subscriptions.

alexhemard replied:
`/subscriptions/*` is not a valid azure IAM scope - the subscription wildcard `*` is a Teleport-specific concept for the discovery config matchers

greedy52 commented on integrations/terraform-modules/teleport/discovery/azure/teleport_provision_token.tf:
```diff
@@ -0,0 +1,58 @@
+
+# Teleport provision token for Azure join
+resource "teleport_provision_token" "azure" {
+  count = local.create ? 1 : 0
```
nit: this is good for now but may not be required in the future if no `vm` matcher

alexhemard committed:
feedback, redux
